### PR TITLE
Condition wait

### PR DESCRIPTION
**Pull request by brookman1**
Pulled the code from leshchenko1979/reretry, modified api.py and test_retry.py so that only the condition_wait changes are added.

Done to comply with request on this link: https://github.com/invl/retry/pull/48#issuecomment-1117635954

Thanks.

### CODE REVIEW

File: retry/api.py [hunk 1]
```diff
@@ -1,84 +1,215 @@
...
+                _log_attempt(tries, show_traceback, logger, _tries, _delay, e)
+
+            if not _tries:
+                raise
+
+            if fail_callback is not None:
+                fail_callback(e)
+
+            with ConditionRelease(condition) as _condition:
+                print('_condtion:', _condition)
```
Comment (leshchenko1979) [inline retry/api.py:57]: Does it really need to print to stdout or did you want to add some logging?

File: retry/api.py [hunk 2]
```diff
@@ -1,84 +1,215 @@
...
+class ConditionRelease(object):
+    def __init__(self, c):
+        self.c = c
+    def __enter__(self):
+        self.c.acquire()
+        return self
+    def wait(self, n):
+        self.c.wait(n)
+    def __exit__(self, exc_type, exc_value, exc_tb):
+        self.c.release()
```
Comment (leshchenko1979) [inline retry/api.py:23]: Can a `contextlib.contextmanager` be used instead? Would it make it simpler?

**Comment by leshchenko1979** [comment]
Merge remote-tracking branch 'origin/master' into pr/brookman1/5
```diff
diff --git a/.github/workflows/python-package.yml b/.github/workflows/python-package.yml
new file mode 100644
index 0000000..c11ecfe
--- /dev/null
+++ b/.github/workflows/python-package.yml
@@ -0,0 +1,42 @@
+# This workflow will install Python dependencies, run tests and lint with a variety of Python versions
+# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+
+name: build
+
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v3
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install flake8 pytest pytest-cov
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+        if [ -f tests/test-requirements.txt ]; then pip install -r tests/test-requirements.txt; fi
+    - name: Lint with flake8
+      run: |
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+    - name: Run tests and collect coverage
+      run: pytest --cov .
+    - name: Upload coverage reports to Codecov with GitHub Action
+      uses: codecov/codecov-action@v3
diff --git a/.github/workflows/python-publish.yml b/.github/workflows/python-publish.yml
new file mode 100644
index 0000000..ec70354
--- /dev/null
+++ b/.github/workflows/python-publish.yml
@@ -0,0 +1,39 @@
+# This workflow will upload a Python Package using Twine when a release is created
+# For more information see: https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions#publishing-to-package-registries
+
+# This workflow uses actions that are not certified by GitHub.
+# They are provided by a third-party and are governed by
+# separate terms of service, privacy policy, and support
+# documentation.
+
+name: Upload Python Package
+
+on:
+  release:
+    types: [published]
+
+permissions:
+  contents: read
+
+jobs:
+  deploy:
+
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: Set up Python
+      uses: actions/setup-python@v3
+      with:
+        python-version: '3.x'
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install build
+    - name: Build package
+      run: python -m build
+    - name: Publish package
+      uses: pypa/gh-action-pypi-publish@27b31702a0e7fc50959f5ad993c78deac1bdfc29
+      with:
+        user: __token__
+        password: ${{ secrets.PYPI_API_TOKEN }}
diff --git a/.gitignore b/.gitignore
index 5e524c3..ec33e5f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -34,3 +34,4 @@ sdist/
 target/
 var/
 __pycache__/
+.vscode/
diff --git a/AUTHORS b/AUTHORS
deleted file mode 100644
index 3ae1286..0000000
--- a/AUTHORS
+++ /dev/null
@@ -1,5 +0,0 @@
-Rémy <rgreinho@users.noreply.github.com>
-Rémy Greinhofer <remy.greinhofer@livelovely.com>
-invlpg <invlpg@gmail.com>
-Richard O'Dwyer <richard@richard.do>
-williara <ray@hudl.com>
diff --git a/ChangeLog b/ChangeLog
deleted file mode 100644
index f537717..0000000
--- a/ChangeLog
+++ /dev/null
@@ -1,139 +0,0 @@
-CHANGES
-=======
-
-0.9.2
------
-
-* Updating requirements.txt to allow for any decorators >=3.4.2
-
-0.9.1
------
-
-* Fix dependency issues with other packages caused by explicit dep verions in requirements
-
-0.9.0
------
-
-* Add AUTHORS and ChangeLog files
-* Packaging the application using PBR
-* Update documentation
-* Add retry_call function
-* Update tox.ini
-* Update requirements
-* Move the tests to the appropriate package
-* Add .gitignore file
-
-0.8.1
------
-
-* v0.8.1
-* Move try/import to compat
-
-0.8.0
------
-
-* v0.8.0
-* dos2unix
-* Add argument jitter for retry()
-* Add argument max_delay for retry()
-* Refactor retry()
-
-0.7.0
------
-
-* v0.7.0
-* Update README.rst
-* retry(): Update docstring
-* retry(): Change default tries to -1
-* Move decorator() to .compat
-* Add test_tries_minus1()
-* Add test_tries_inf()
-* Mock time.sleep in test case
-* Refactor retry()
-
-0.6.0
------
-
-* v0.6.0
-* Fix inaccurate attempt counter
-* logger is now optional
-* Extract logging_logger
-* Make decorator module optional
-
-0.5.0
------
-
-* v0.5.0
-* Update README.rst
-* Add badges
-* Configurable logger
-* Update classifiers
-* Remove .hgignore and .hgtags
-* Faster test
-* Support Python 3.4
-* Add tox.ini
-* Extract retry/api.py
-* Require pytest
-* Add test_retry.py
-* Added tag 0.4.2 for changeset 315f5f1229f6
-
-0.4.2
------
-
-* Version 0.4.2
-* python2.6 support
-* README.rst: Add installation
-* Add classifiers
-* Add LICENSE
-* Add requirements.txt
-* Fix rST h1 h2 for README.rst
-* Add url
-* Add README.rst
-* Ignore *.egg-info
-* Ignore .env
-* Ignore .ropeproject
-* Ignore .git
-* Added tag 0.4.1 for changeset 2765d4be1b4d
-
-0.4.1
------
-
-* Version 0.4.1
-* Add license
-* Add long_description
-* Add docstring for retry()
-* Added tag 0.4.0 for changeset e053cae4b105
-
-0.4.0
------
-
-* Version 0.4.0
-* Convert to a package
-* backoff defaults to 1
-* Refactor: tries defaults to inf instead of None
-* Added tag 0.3.0 for changeset 60960fe42471
-
-0.3.0
------
-
-* Version 0.3.0
-* delay defaults to 0
-* Use logging
-* tries can be None, for retry forever
-* Add ifmain
-* Refactor
-* Added tag 0.2.0 for changeset 87b6dd2bc30b
-
-0.2.0
------
-
-* Fix typo
-* Refactor
-* Add description
-* Use decorator
-* Added tag 0.1.0 for changeset bf83184dd128
-
-0.1.0
------
-
-* init
diff --git a/LICENSE b/LICENSE
index 3a2cc32..7ed61ab 100644
--- a/LICENSE
+++ b/LICENSE
@@ -1,4 +1,4 @@
-Copyright 2014 invl
+Copyright 2014 invl, 2022 leshchenko1979
 
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
diff --git a/README.md b/README.md
new file mode 100644
index 0000000..9196104
--- /dev/null
+++ b/README.md
@@ -0,0 +1,147 @@
+# reretry
+
+![](https://github.com/leshchenko1979/reretry/workflows/build/badge.svg)
+[![codecov](https://codecov.io/gh/leshchenko1979/reretry/branch/master/graph/badge.svg)](https://codecov.io/gh/leshchenko1979/reretry)
+![](https://img.shields.io/pypi/dm/reretry.svg)
+
+An easy to use retry decorator.
+
+This package is a fork from the [`retry`](https://github.com/invl/retry) package, but with some of added community-sourced features.
+
+
+## Features
+
+From original `retry`:
+- Retry on specific exceptions.
+- Set a maximum number of retries.
+- Set a delay between retries.
+- Set a maximum delay between retries.
+- Set backoff and jitter parameters.
+- Use a custom logger.
+- No external dependencies (stdlib only).
+- (Optionally) Preserve function signatures (`pip install decorator`).
+
+New features in `reretry`:
+- Log traceback of an error that lead to a failed attempt.
+- Call a custom callback after each failed attempt.
+- Can be used with async functions.
+
+
+## Installation
+
+```bash
+$ pip install reretry
+```
+
+## API
+### The @retry decorator
+
+#### Usage
+`@retry(exceptions=Exception, tries=-1, delay=0, max_delay=None, backoff=1, jitter=0, show_traceback=False, logger=logging_logger, fail_callback=None)`
+
+#### Arguments
+- `exceptions`: An exception or a tuple of exceptions to catch. Default: Exception.
+
+- `tries`: The maximum number of attempts. default: -1 (infinite).
+
+- `delay`: Initial delay between attempts (in seconds). default: 0.
+
+- `max_delay`: The maximum value of delay (in seconds). default: None (no limit).
+
+- `backoff`: Multiplier applied to delay between attempts. default: 1 (no backoff).
+
+- `jitter`: Extra seconds added to delay between attempts. default: 0. Fixed if a number, random if a range tuple (min, max).
+
+- `show_traceback`: Print traceback before retrying (Python3 only). default: False.
+
+- `logger`: `logger.warning(fmt, error, delay)` will be called on failed attempts. default: retry.logging_logger. if None, logging is disabled.
+
+- `fail_callback`: `fail_callback(e)` will be called after failed attempts.
+
+
+#### Examples
+```python
+from reretry import retry
+
+@retry()
+def make_trouble():
+    '''Retry until succeeds'''
+
+@retry()
+async def async_make_trouble():
+    '''Retry an async function until it succeeds'''
+
+@retry(ZeroDivisionError, tries=3, delay=2)
+def make_trouble():
+    '''Retry on ZeroDivisionError, raise error after 3 attempts,
+    sleep 2 seconds between attempts.'''
+
+@retry((ValueError, TypeError), delay=1, backoff=2)
+def make_trouble():
+    '''Retry on ValueError or TypeError, sleep 1, 2, 4, 8, ... seconds between attempts.'''
+
+@retry((ValueError, TypeError), delay=1, backoff=2, max_delay=4)
+def make_trouble():
+    '''Retry on ValueError or TypeError, sleep 1, 2, 4, 4, ... seconds between attempts.'''
+
+@retry(ValueError, delay=1, jitter=1)
+def make_trouble():
+    '''Retry on ValueError, sleep 1, 2, 3, 4, ... seconds between attempts.'''
+
+def callback(e: Exception):
+    '''Print error message'''
+    print(e)
+
+@retry(ValueError, fail_callback=callback):
+def make_trouble():
+    '''Retry on ValueError, between attempts call callback(e)
+    (where e is the Exception raised).'''
+
+# If you enable logging, you can get warnings like 'ValueError, retrying in
+# 1 seconds'
+if __name__ == '__main__':
+    import logging
+    logging.basicConfig()
+    make_trouble()
+```
+
+### The `retry_call` function
+Calls a function and re-executes it if it failed.
+
+This is very similar to the decorator, except that it takes a function and its arguments as parameters. The use case behind it is to be able to dynamically adjust the retry arguments.
+
+#### Usage
+`retry_call(f, fargs=None, fkwargs=None, exceptions=Exception, tries=-1, delay=0, max_delay=None, backoff=1, jitter=0, show_traceback=False, logger=logging_logger, fail_callback=None)`
+
+#### Example
+```python
+import requests
+
+from reretry.api import retry_call
+
+
+def make_trouble(service, info=None):
+    if not info:
+        info = ''
+    r = requests.get(service + info)
+    return r.text
+
+
+def what_is_my_ip(approach=None):
+    if approach == "optimistic":
+        tries = 1
+    elif approach == "conservative":
+        tries = 3
+    else:
+        # skeptical
+        tries = -1
+    result = retry_call(
+        make_trouble,
+        fargs=["http://ipinfo.io/"],
+        fkwargs={"info": "ip"},
+        tries=tries
+    )
+    print(result)
+
+what_is_my_ip("conservative")
+```
* Unmerged path README.rst
diff --git a/codecov.yaml b/codecov.yaml
new file mode 100644
index 0000000..a0ee01d
--- /dev/null
+++ b/codecov.yaml
@@ -0,0 +1,3 @@
+ignore:
+  - "tests"
+  - "setup.py"
diff --git a/requirements.txt b/requirements.txt
deleted file mode 100644
index 5c13ceb..0000000
--- a/requirements.txt
+++ /dev/null
@@ -1,2 +0,0 @@
-decorator>=3.4.2
-py>=1.4.26,<2.0.0
diff --git a/reretry/__init__.py b/reretry/__init__.py
new file mode 100644
index 0000000..805113a
--- /dev/null
+++ b/reretry/__init__.py
@@ -0,0 +1,3 @@
+__all__ = ["retry", "retry_call"]
+
+from .api import retry, retry_call
diff --git a/reretry/api.py b/reretry/api.py
new file mode 100644
index 0000000..7cf3170
--- /dev/null
+++ b/reretry/api.py
@@ -0,0 +1,229 @@
+import asyncio
+import logging
+import random
+import time
+import traceback
+from functools import partial
+
+import inspect
+
+from .compat import decorator
+
+logging_logger = logging.getLogger(__name__)
+
+
+def __retry_internal(
+    f,
+    exceptions=Exception,
+    tries=-1,
+    delay=0,
+    max_delay=None,
+    backoff=1,
+    jitter=0,
+    show_traceback=False,
+    logger=logging_logger,
+    fail_callback=None,
+):
+    _tries, _delay = tries, delay
+
+    while _tries:
+        try:
+            return f()
+
+        except exceptions as e:
+            _tries -= 1
+
+            if logger:
+                _log_attempt(tries, show_traceback, logger, _tries, _delay, e)
+
+            if not _tries:
+                raise
+
+            if fail_callback is not None:
+                fail_callback(e)
+
+            time.sleep(_delay)
+
+            _delay = _new_delay(max_delay, backoff, jitter, _delay)
+
+
+async def __retry_internal_async(
+    f,
+    exceptions=Exception,
+    tries=-1,
+    delay=0,
+    max_delay=None,
+    backoff=1,
+    jitter=0,
+    show_traceback=False,
+    logger=logging_logger,
+    fail_callback=None,
+):
+    _tries, _delay = tries, delay
+
+    while _tries:
+        try:
+            return await f()
+
+        except exceptions as e:
+            _tries -= 1
+
+            if logger:
+                _log_attempt(tries, show_traceback, logger, _tries, _delay, e)
+
+            if not _tries:
+                raise
+
+            if fail_callback is not None:
+                await fail_callback(e)
+
+            await asyncio.sleep(_delay)
+
+            _delay = _new_delay(max_delay, backoff, jitter, _delay)
+
+
+def _log_attempt(tries, show_traceback, logger, _tries, _delay, e):
+    if _tries:
+        if show_traceback:
+            tb_str = "".join(traceback.format_exception(None, e, e.__traceback__))
+            logger.warning(tb_str)
+
+        logger.warning(
+            "%s, attempt %s/%s failed - retrying in %s seconds...",
+            e,
+            tries - _tries,
+            tries,
+            _delay,
+        )
+
+    elif tries > 1:
+        logger.warning(
+            "%s, attempt %s/%s failed - giving up!", e, tries - _tries, tries
+        )
+
+
+def _new_delay(max_delay, backoff, jitter, _delay):
+    _delay *= backoff
+    _delay += random.uniform(*jitter) if isinstance(jitter, tuple) else jitter
+
+    if max_delay is not None:
+        _delay = min(_delay, max_delay)
+
+    return _delay
+
+
+def _is_async(f):
+    return asyncio.iscoroutinefunction(f) and not inspect.isgeneratorfunction(f)
+
+
+def _get_internal_function(f):
+    return __retry_internal_async if _is_async(f) else __retry_internal
+
+
+def _check_params(f, show_traceback=False, logger=logging_logger, fail_callback=None):
+    assert not show_traceback or logger is not None, "`show_traceback` needs `logger`"
+
+    assert not fail_callback or (
+        (_is_async(f) and _is_async(fail_callback))
+        or (not _is_async(f) and not _is_async(fail_callback))
+    ), "If the retried function is async, fail_callback needs to be async as well or vice versa"
+
+
+def retry(
+    exceptions=Exception,
+    tries=-1,
+    delay=0,
+    max_delay=None,
+    backoff=1,
+    jitter=0,
+    show_traceback=False,
+    logger=logging_logger,
+    fail_callback=None,
+):
+    """Returns a retry decorator.
+
+    :param exceptions: an exception or a tuple of exceptions to catch. default: Exception.
+    :param tries: the maximum number of attempts. default: -1 (infinite).
+    :param delay: initial delay between attempts (in seconds). default: 0.
+    :param max_delay: the maximum value of delay (in seconds). default: None (no limit).
+    :param backoff: multiplier applied to delay between attempts. default: 1 (no backoff).
+    :param jitter: extra seconds added to delay between attempts. default: 0.
+                   fixed if a number, random if a range tuple (min, max)
+    :param show_traceback: if True, the traceback of the exception will be logged.
+    :param logger: logger.warning(fmt, error, delay) will be called on failed attempts.
+                   default: retry.logging_logger. if None, logging is disabled.
+    :param fail_callback: fail_callback(e) will be called on failed attempts.
+    :returns: a retry decorator.
+    """
+
+    @decorator
+    def retry_decorator(f, *fargs, **fkwargs):
+        return retry_call(
+            f,
+            fargs,
+            fkwargs,
+            exceptions,
+            tries,
+            delay,
+            max_delay,
+            backoff,
+            jitter,
+            show_traceback,
+            logger,
+            fail_callback,
+        )
+
+    return retry_decorator
+
+
+def retry_call(
+    f,
+    fargs=None,
+    fkwargs=None,
+    exceptions=Exception,
+    tries=-1,
+    delay=0,
+    max_delay=None,
+    backoff=1,
+    jitter=0,
+    show_traceback=False,
+    logger=logging_logger,
+    fail_callback=None,
+):
+    """
+    Calls a function and re-executes it if it failed.
+
+    :param f: the function to execute.
+    :param fargs: the positional arguments of the function to execute.
+    :param fkwargs: the named arguments of the function to execute.
+    :param exceptions: an exception or a tuple of exceptions to catch. default: Exception.
+    :param tries: the maximum number of attempts. default: -1 (infinite).
+    :param delay: initial delay between attempts (in seconds). default: 0.
+    :param max_delay: the maximum value of delay (in seconds). default: None (no limit).
+    :param backoff: multiplier applied to delay between attempts. default: 1 (no backoff).
+    :param jitter: extra seconds added to delay between attempts. default: 0.
+                   fixed if a number, random if a range tuple (min, max)
+    :param show_traceback: if True, the traceback of the exception will be logged.
+    :param logger: logger.warning(fmt, error, delay) will be called on failed attempts.
+                   default: retry.logging_logger. if None, logging is disabled.
+    :param fail_callback: fail_callback(e) will be called on failed attempts.
+    :returns: the result of the f function.
+    """
+    args = fargs or list()
+    kwargs = fkwargs or dict()
+
+    _check_params(f, show_traceback, logger, fail_callback)
+
+    func = _get_internal_function(f)
+    return func(
+        partial(f, *args, **kwargs),
+        exceptions,
+        tries,
+        delay,
+        max_delay,
+        backoff,
+        jitter,
+        show_traceback,
+        logger,
+        fail_callback,
+    )
diff --git a/reretry/compat.py b/reretry/compat.py
new file mode 100644
index 0000000..b26b05c
--- /dev/null
+++ b/reretry/compat.py
@@ -0,0 +1,24 @@
+import functools
+
+
+def simple_decorator(caller):
+    """Turns caller into a decorator.
+    Unlike decorator module, function signature is not preserved.
+
+    :param caller: caller(f, *args, **kwargs)
+    """
+
+    def decor(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            return caller(f, *args, **kwargs)
+
+        return wrapper
+
+    return decor
+
+
+try:
+    from decorator import decorator
+except ImportError:
+    decorator = simple_decorator
diff --git a/retry/__init__.py b/retry/__init__.py
deleted file mode 100644
index 831b21e..0000000
--- a/retry/__init__.py
+++ /dev/null
@@ -1,11 +0,0 @@
-__all__ = ['retry', 'retry_call']
-
-import logging
-
-from .api import retry, retry_call
-from .compat import NullHandler
-
-
-# Set default logging handler to avoid "No handler found" warnings.
-log = logging.getLogger(__name__)
-log.addHandler(NullHandler())
* Unmerged path retry/api.py
diff --git a/retry/compat.py b/retry/compat.py
deleted file mode 100644
index 4a4a06d..0000000
--- a/retry/compat.py
+++ /dev/null
@@ -1,28 +0,0 @@
-import functools
-import logging
-
-
-try:
-    from decorator import decorator
-except ImportError:
-    def decorator(caller):
-        """ Turns caller into a decorator.
-        Unlike decorator module, function signature is not preserved.
-
-        :param caller: caller(f, *args, **kwargs)
-        """
-        def decor(f):
-            @functools.wraps(f)
-            def wrapper(*args, **kwargs):
-                return caller(f, *args, **kwargs)
-            return wrapper
-        return decor
-
-
-try:  # Python 2.7+
-    from logging import NullHandler
-except ImportError:
-    class NullHandler(logging.Handler):
-
-        def emit(self, record):
-            pass
diff --git a/setup.cfg b/setup.cfg
index d723671..ed1c6d6 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -1,13 +1,14 @@
 [metadata]
-name = retry
-version = 0.9.3
-author = invl
-author-email = invlpg@gmail.com
-summary = Easy to use retry decorator.
+name = reretry
+version = 0.11.2
+author = leshchenko1979
+author_email = leshchenko1979@gmail.com
+summary = An easy to use, but functional decorator for retrying on exceptions.
 license = Apache License 2.0
-description-file = README.rst
-home-page = https://github.com/invl/retry
-requires-python = >=2.6
+description_file = README.md
+long_description_content_type = text/x-rst
+home_page = https://github.com/leshchenko1979/reretry
+requires_python = >=3.6
 classifier =
     Development Status :: 4 - Beta
     Intended Audience :: Developers
@@ -15,12 +16,13 @@ classifier =
     Natural Language :: English
     Operating System :: OS Independent
     Programming Language :: Python
-    Programming Language :: Python :: 2.6
-    Programming Language :: Python :: 2.7
-    Programming Language :: Python :: 3
-    Programming Language :: Python :: 3.4
-    Programming Language :: Python :: Implementation :: PyPy
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.8
+    Programming Language :: Python :: 3.9
+    Programming Language :: Python :: 3.10
+    Programming Language :: Python :: Implementation :: CPython
     Topic :: Software Development
 
-[wheel]
+[bdist_wheel]
 universal = 1
diff --git a/test-requirements.txt b/test-requirements.txt
deleted file mode 100644
index b84cb24..0000000
--- a/test-requirements.txt
+++ /dev/null
@@ -1,5 +0,0 @@
-mock
-pbr
-pytest
-tox
-wheel
diff --git a/tests/__init__.py b/tests/__init__.py
new file mode 100644
index 0000000..e69de29
diff --git a/tests/test-requirements.txt b/tests/test-requirements.txt
new file mode 100644
index 0000000..2d73dba
--- /dev/null
+++ b/tests/test-requirements.txt
@@ -0,0 +1 @@
+pytest-asyncio
diff --git a/tests/test_async.py b/tests/test_async.py
new file mode 100644
index 0000000..9234a69
--- /dev/null
+++ b/tests/test_async.py
@@ -0,0 +1,84 @@
+from unittest.mock import MagicMock
+
+import asyncio
+
+import pytest
+from reretry.api import _is_async, retry, retry_call
+
+
+
+def test_is_async():
+    async def async_func():
+        pass
+
+    def non_async_func():
+        pass
+
+    def generator():
+        yield
+
+    assert _is_async(async_func)
+    assert not _is_async(non_async_func)
+    assert not _is_async(generator)
+    assert not _is_async(generator())
+    assert not _is_async(MagicMock(spec=non_async_func, return_value=-1))
+
+
+@pytest.mark.asyncio
+async def test_async():
+    attempts = 1
+    raised = False
+
+    @retry(tries=2)
+    async def f():
+        await asyncio.sleep(0.01)
+        nonlocal attempts, raised
+        if attempts:
+            raised = True
+            attempts -= 1
+            raise RuntimeError
+        return True
+
+    assert await f()
+    assert raised
+    assert attempts == 0
+
+
+@pytest.mark.asyncio
+async def test_async_fail_and_callback():
+    cb_called = False
+
+    async def cb(exception: Exception):
+        nonlocal cb_called
+        cb_called = True
+        assert exception.args[0] == 1
+
+    attempts = 3
+
+    @retry(tries=2, fail_callback=cb)
+    async def f():
+        await asyncio.sleep(0.01)
+        nonlocal attempts
+        if attempts:
+            attempts -= 1
+            raise RuntimeError(1)
+        return True
+
+    with pytest.raises(RuntimeError):
+        await f()
+
+    assert cb_called
+
+
+def test_check_params():
+    with pytest.raises(AssertionError):
+        retry_call(lambda: None, show_traceback=True, logger=None)
+
+    async def async_func():
+        pass
+
+    def non_async_func():
+        pass
+
+    with pytest.raises(AssertionError):
+        retry_call(async_func, fail_callback=non_async_func)
* Unmerged path tests/test_retry.py
diff --git a/tox.ini b/tox.ini
deleted file mode 100644
index 2f183e5..0000000
--- a/tox.ini
+++ /dev/null
@@ -1,23 +0,0 @@
-# Tox (http://tox.testrun.org/) is a tool for running tests
-# in multiple virtualenvs. This configuration file will run the
-# test suite on all supported python versions. To use it, "pip install tox"
-# and then run "tox" from this directory.
-
-[tox]
-envlist = flake8, py26, py27, py34, py35, pypy
-
-[flake8]
-max-line-length = 120
-exclude = *.cfg,*.egg,*.ini,*.log,*tests*,*.txt,*.xml,.tox,.venv,AUTHORS,build,ChangeLog,dist,doc,test-requirements.txt,src
-format = pylint
-
-[testenv]
-commands = py.test
-deps=
-    -rrequirements.txt
-    -rtest-requirements.txt
-
-[testenv:flake8]
-basepython = python3.4
-commands = flake8
-deps = flake8

```

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]
Another questions -- you propose that all the old tests should now have threads in them. Shouldn't you create a new set of tests for your new feature?

File: retry/api.py (hunk 1 above)
Comment (brookman1) [reply retry/api.py]: removing.

File: retry/api.py (hunk 2 above)
Comment (brookman1) [reply retry/api.py]: Will change the context manager class, but contextmanager decorator alone is not something that seems fullproof enough.  This context manager class will make code easier to read.

**Review by leshchenko1979 — CHANGES_REQUESTED** [review]
Sorry, it still hasn't been rebased on the latest commint in `reretry:mater`. It contains marge conflits which I am afriad I am unable to resolve.

**Comment by brookman1** [comment]
I'd like to withdraw the request, this link explain why it is not needed:
https://stackoverflow.com/questions/92928/time-sleep-sleeps-thread-or-process

Thanks.

**Pull request closed without merging**